zcash_client_sqlite/src/wallet/scanning.rs

```diff
@@ -0,0 +1,1089 @@
+                        // extend the left-hand branch
+                        Self::from_insert(left, right, to_insert, Insert::Left)
+                    }
+                }
+            }
+        }
+    }
+
+    fn into_vec(self) -> Vec<ScanRange> {
+        fn go(acc: &mut Vec<ScanRange>, tree: SpanningTree) {
+            match tree {
+                SpanningTree::Leaf(entry) => {
+                    if let Some(top) = acc.pop() {
+                        match join_nonoverlapping(top, entry) {
+                            Joined::One(merged) => acc.push(merged),
+                            Joined::Two(l, r) => {
+                                acc.push(l);
+                                acc.push(r);
+                            }
+                            _ => unreachable!(),
+                        }
+                    } else {
+                        acc.push(entry);
+                    }
+                }
+                SpanningTree::Parent { left, right, .. } => {
+                    go(acc, *left);
+                    go(acc, *right);
+                }
+            }
+        }
+
+        let mut acc = vec![];
+        go(&mut acc, self);
+        acc
+    }
+}
+
+pub(crate) fn insert_queue_entries<'a>(
+    conn: &rusqlite::Connection,
+    entries: impl Iterator<Item = &'a ScanRange>,
+) -> Result<(), rusqlite::Error> {
+    let mut stmt = conn.prepare_cached(
+        "INSERT INTO scan_queue (block_range_start, block_range_end, priority)
+        VALUES (:block_range_start, :block_range_end, :priority)",
+    )?;
+
+    for entry in entries {
+        trace!("Inserting queue entry {}", entry);
+        if !entry.is_empty() {
+            stmt.execute(named_params![
+                ":block_range_start": u32::from(entry.block_range().start),
+                ":block_range_end": u32::from(entry.block_range().end),
+                ":priority": priority_code(&entry.priority())
+            ])?;
+        }
+    }
+
+    Ok(())
+}
+
+pub(crate) fn replace_queue_entries(
+    conn: &rusqlite::Transaction<'_>,
+    query_range: &Range<BlockHeight>,
+    entries: impl Iterator<Item = ScanRange>,
+) -> Result<(), SqliteClientError> {
+    let (to_create, to_delete_ends) = {
+        let mut suggested_stmt = conn.prepare_cached(
+            "SELECT block_range_start, block_range_end, priority
+            FROM scan_queue
+            WHERE (
+                -- the start is contained within the range
+                :start >= block_range_start
+                AND :start < block_range_end
+            )
+            OR (
+                -- the end is contained within the range
+                :end > block_range_start
+                AND :end <= block_range_end
+            )
+            OR (
+                -- start..end contains the entire range
+                block_range_start >= :start
+                AND block_range_end <= :end
+            )
+            ORDER BY block_range_end",
+        )?;
```

Comment:
An empty `query_range` might return a match even though its intersection with every range is empty. I think this doesn't matter.